yktrace/src/sir.rs

```diff
@@ -1,92 +1,148 @@
-        // memory.
-        let mut bodies = HashMap::new();
-        let mut types = HashMap::new();
-        for sec in object.sections() {
+        fn read_hdr(curs: &mut Cursor<&[u8]>) -> SirHeader {
+            let mut dec = Decoder::from(curs);
+            if let Ok(Some(Pack::Header(hdr))) = dec.next() {
+                hdr
+            } else {
+                panic!("failed to read sir header");
+            }
+        }
+
+        let mut hdrs = FxHashMap::default();
+        let exe_obj = object::File::parse(&*mmap).unwrap();
+        for sec in exe_obj.sections() {
             if sec.name().unwrap().starts_with(".yksir_") {
                 let mut curs = Cursor::new(sec.data().unwrap());
-                let mut dec = Decoder::from(&mut curs);
-
-                while let Some(pack) = dec.next().unwrap() {
-                    match pack {
-                        Pack::Body(body) => {
-                            // Due to the way Rust compiles stuff, duplicates may exist. Where
-                            // duplicates exist, the functions will be identical, but may have
-                            // different (but equivalent) types. This is because types too may be
-                            // duplicated (for example in a different crate).
-                            bodies
-                                .entry(body.symbol_name.clone())
-                                .or_insert_with(|| body);
-                        }
-                        Pack::Types(ts) => {
-                            let old = types.insert(ts.cgu_hash, ts.types);
-                            debug_assert!(old.is_none()); // There's one `Types` pack per codegen unit.
-                        }
-                    }
-                }
+                let hdr = read_hdr(&mut curs);
+                let hdr_size = usize::try_from(curs.seek(SeekFrom::Current(0)).unwrap()).unwrap();
+                hdrs.insert(
+                    hdr.cgu_hash,
+                    (sec.name().unwrap().to_owned(), hdr, hdr_size)
+                );
             }
         }
+        Ok(Self { hdrs, exe_obj })
+    }
 
-        Ok(Sir { bodies, types })
+    fn cursor_for_section(&self, sec_name: &str) -> Cursor<&[u8]> {
+        let sec = self.exe_obj.section_by_name(&sec_name).unwrap();
+        Cursor::new(sec.data().unwrap())
     }
-}
 
-impl Display for Sir {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for body in self.bodies.values() {
-            writeln!(f, "{}", body)?;
+    /// Decode a type in a named section, at an absolute offset from the beginning of that section.
+    fn decode_ty(&self, sec_name: &str, off: SirOffset) -> ykpack::Ty {
+        let mut curs = self.cursor_for_section(&sec_name);
+        curs.seek(SeekFrom::Start(u64::try_from(off).unwrap()))
+            .unwrap();
+        let mut dec = Decoder::from(&mut curs);
+        if let Ok(Some(Pack::Type(t))) = dec.next() {
```

Comment:
What does `next` return here? I wonder if this is just a long version of `dec.next().unwrap().unwrap().0`?

Reply:
Good point.

Reply:
I couldn't simplify it quite to that degree, as we need to destructure the header variant of the pack.

If this really bothers us, we can make the header independent of `Pack` and make this new type serde compatible.

I'm not sure it's that' big of an issue though?

Reply:
It's really whatever you think is easiest!